### PR TITLE
stageblock | appendblock | putpages StructuredMessage

### PR DESCRIPTION
**Pull request by jaschrep-msft**
Use structured message in blobservice 1:1 puts when no precalculated checksum is provided.

Tests branched for datalake and shares. Will be reunited once implementation is added to datalake append and share put.

### CODE REVIEW

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageVersionExtensions.cs [hunk 1]
```diff
@@ -42,7 +42,7 @@ internal static class StorageVersionExtensions
         /// </summary>
         public const ServiceVersion LatestVersion =
 #if BlobSDK || QueueSDK || FileSDK || DataLakeSDK || ChangeFeedSDK || DataMovementSDK || BlobDataMovementSDK || ShareDataMovementSDK
-            ServiceVersion.V2024_02_04;
+            ServiceVersion.V2025_01_05;
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/StorageVersionExtensions.cs:45]: We probably need to revert this line, the recorded SAS tests will probably fail now.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TransferValidationTestBase.cs [hunk 2]
```diff
@@ -32,7 +32,7 @@ public abstract class TransferValidationTestBase<TServiceClient, TContainerClien
             bool async,
             string generatedResourceNamePrefix = default,
             RecordedTestMode? mode = null)
-            : base(async, mode)
+            : base(async, RecordedTestMode.Record)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/TransferValidationTestBase.cs:35]: Revert

File: sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs [hunk 3]
```diff
@@ -1241,14 +1241,33 @@ public new AppendBlobClient WithEncryptionScope(string encryptionScope)
...
+                    string structuredBodyType = null;
+                    if (validationOptions != null &&
+                        validationOptions.ChecksumAlgorithm.ResolveAuto() == StorageChecksumAlgorithm.StorageCrc64 &&
+                        validationOptions.PrecalculatedChecksum.IsEmpty &&
+                        ClientSideEncryption == null) // don't allow feature combination
+                    {
+                        // report progress in terms of caller bytes, not encoded bytes
+                        structuredContentLength = contentLength;
+                        contentLength = (content?.Length - content?.Position) ?? 0;
+                        structuredBodyType = "XSM/1.0; properties=crc64";
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:1256]: We should put this in Constants in the Common package.

File: sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs [hunk 4]
```diff
@@ -1241,14 +1241,33 @@ public new AppendBlobClient WithEncryptionScope(string encryptionScope)
...
+                        content = content.WithNoDispose().WithProgress(progressHandler);
+                        content = new StructuredMessageEncodingStream(content, 4 * Constants.MB, StructuredMessage.Flags.StorageCrc64);
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:1258]: Same with the `4 * Constants.MB`

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs [hunk 5]
```diff
@@ -1305,22 +1305,41 @@ public new BlockBlobClient WithEncryptionScope(string encryptionScope)
...
+                    string structuredBodyType = null;
+                    if (validationOptions != null &&
+                        validationOptions.ChecksumAlgorithm.ResolveAuto() == StorageChecksumAlgorithm.StorageCrc64 &&
+                        validationOptions.PrecalculatedChecksum.IsEmpty &&
+                        ClientSideEncryption == null) // don't allow feature combination
+                    {
+                        // report progress in terms of caller bytes, not encoded bytes
+                        structuredContentLength = contentLength;
+                        contentLength = (content?.Length - content?.Position) ?? 0;
+                        structuredBodyType = "XSM/1.0; properties=crc64";
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:1320]: Constants

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs [hunk 6]
```diff
@@ -1305,22 +1305,41 @@ public new BlockBlobClient WithEncryptionScope(string encryptionScope)
...
+                        content = content.WithNoDispose().WithProgress(progressHandler);
+                        content = new StructuredMessageEncodingStream(content, 4 * Constants.MB, StructuredMessage.Flags.StorageCrc64);
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:1322]: 4 * Constants.MB to Constants.

File: sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs [hunk 7]
```diff
@@ -1363,15 +1363,36 @@ public new PageBlobClient WithEncryptionScope(string encryptionScope)
...
+                    if (validationOptions != null &&
+                        validationOptions.ChecksumAlgorithm.ResolveAuto() == StorageChecksumAlgorithm.StorageCrc64 &&
+                        validationOptions.PrecalculatedChecksum.IsEmpty &&
+                        ClientSideEncryption == null) // don't allow feature combination
+                    {
+                        // report progress in terms of caller bytes, not encoded bytes
+                        structuredContentLength = contentLength;
+                        contentLength = (content?.Length - content?.Position) ?? 0;
+                        range = new HttpRange(offset, (content?.Length - content?.Position) ?? null);
+                        structuredBodyType = "XSM/1.0; properties=crc64";
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:1380]: Constants

File: sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs [hunk 8]
```diff
@@ -1363,15 +1363,36 @@ public new PageBlobClient WithEncryptionScope(string encryptionScope)
...
+                        content = content.WithNoDispose().WithProgress(progressHandler);
+                        content = new StructuredMessageEncodingStream(content, 4 * Constants.MB, StructuredMessage.Flags.StorageCrc64);
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:1382]: Same with 4 * Constants.MB

**Review by seanmcc-msft — COMMENTED** [review]
Very minor comments, looking good!

**Comment by seanmcc-msft** [comment]
I think the issue with the analyze step is we need to run `eng\scripts\Export-API.ps1 storage`

**Comment by seanmcc-msft** [comment]
@jaschrep-msft Looks like CI is complaining about a test recording issue, I can help debug tomorrow.

**Review by jalauzon-msft — APPROVED** [review]

**Pull request merged**